File: phase-1/refactoring/elixir/test/assertion_action_service_test.exs
defmodule Refactoring.AssertionActionServiceTest do
  use ExUnit.Case
  alias Refactoring.AssertionActionService

  test "valid assertion with claims required" do
    request = %{is_claims_required: true}
    claims_infos = %{email: "test@test.fr", username: "test", first_name: "test", last_name: "test"}
    response = AssertionActionService.get_assertion_action_response(request, claims_infos, "test@test.fr")
    assert response.status == :success
  end

  test "invalid assertion with claims required" do
    request = %{is_claims_required: true}
    claims_infos = %{email: "", username: "", first_name: "", last_name: ""}
    response = AssertionActionService.get_assertion_action_response(request, claims_infos, "test@test.fr")
    assert response.status == :error
  end

  test "valid assertion with claims infos empty but claims not required" do
    request = %{is_claims_required: false}
    claims_infos = %{email: "", username: "", first_name: "", last_name: ""}
    response = AssertionActionService.get_assertion_action_response(request, claims_infos, "test@test.fr")
    assert response.status == :success
  end
end

File: phase-1/refactoring/java/src/test/java/AssertionActionServiceTest.java
import org.junit.jupiter.api.Test;
import com.example.refactoring.*;

import static org.junit.jupiter.api.Assertions.*;

public class AssertionActionServiceTest {
    private final AssertionActionService service = new AssertionActionService();

    @Test
    public void testValidAssertion_WithClaimsRequired() {
        AssertionConsumerServiceRequest request = new AssertionConsumerServiceRequest();
        request.setClaimsRequired(true);

        ClaimsInfos claimsInfos = new ClaimsInfos();
        claimsInfos.setEmail("test@test.fr");
        claimsInfos.setUserName("test");
        claimsInfos.setFirstName("test");
        claimsInfos.setLastName("test");

        AssertionActionResponse response = service.getAssertionActionResponse(request, claimsInfos, "test@test.fr");
        assertEquals(AssertionActionStatus.SUCCESS, response.getStatus());
    }

    @Test
    public void testInvalidAssertion_WithClaimsRequired() {
        AssertionConsumerServiceRequest request = new AssertionConsumerServiceRequest();
        request.setClaimsRequired(true);

        ClaimsInfos claimsInfos = new ClaimsInfos();
        claimsInfos.setEmail("");
        claimsInfos.setUserName("");
        claimsInfos.setFirstName("");
        claimsInfos.setLastName("");

        AssertionActionResponse response = service.getAssertionActionResponse(request, claimsInfos, "test@test.fr");
        assertEquals(AssertionActionStatus.ERROR, response.getStatus());
    }

    @Test
    public void testValidAssertion_WithClaimsInfos_Empty_But_ClaimsNotRequired() {
        AssertionConsumerServiceRequest request = new AssertionConsumerServiceRequest();
        request.setClaimsRequired(false);

        ClaimsInfos claimsInfos = new ClaimsInfos();
        claimsInfos.setEmail("");
        claimsInfos.setUserName("");
        claimsInfos.setFirstName("");
        claimsInfos.setLastName("");

        AssertionActionResponse response = service.getAssertionActionResponse(request, claimsInfos, "test@test.fr");
        assertEquals(AssertionActionStatus.SUCCESS, response.getStatus());
    }

    @Test
    public void testInvalidAssertion_WithSpecificClaimMissing() {
        AssertionConsumerServiceRequest request = new AssertionConsumerServiceRequest();
        request.setClaimsRequired(true);

        ClaimsInfos claimsInfos = new ClaimsInfos();
        claimsInfos.setEmail("");
        claimsInfos.setUserName("username");
        claimsInfos.setFirstName("firstname");
        claimsInfos.setLastName("lastname");

        AssertionActionResponse response = service.getAssertionActionResponse(request, claimsInfos, "test@test.com");
        assertEquals(AssertionActionStatus.ERROR, response.getStatus());
    }
}